### PR TITLE
Part2

### PR DESCRIPTION
**Pull request by IdanTo1**
hope for the best, prepare for the worst

### CODE REVIEW

File: part2/mermaid.ts [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+                                            )
+                );
+};
+
+const parseSexpValue = (exp: SExpValue, head: M.NodeRef, label?: M.EdgeLabel): Result<M.Edge[]> =>
+    S.isNumber(exp) ?  makeOk([M.makeEdge(head, M.makeNodeDecl(makeCountGen("number"), `"number(${exp.toString()})"`), label)]) :
+    S.isBoolean(exp) ?  makeOk([M.makeEdge(head, M.makeNodeDecl(makeCountGen("boolean"), `"boolean(${exp.toString()})"`), label)]) :
+    S.isString(exp) ?  makeOk([M.makeEdge(head, M.makeNodeDecl(makeCountGen("string"), `"string(${exp})"`), label)]) :
+    L4.isPrimOp(exp) ?  makeOk([M.makeEdge(head, M.makeNodeDecl(makeCountGen("PrimOp"), `"PrimOp(${exp.op})"`), label)]) :
+    isSymbolSExp(exp) ?  makeOk([M.makeEdge(head, M.makeNodeDecl(makeCountGen("SymbolExp"), `"SymbolExp(${exp.val})"`), label)]) :
```
Comment (IdanTo1) [inline part2/mermaid.ts:190]: should be SymbolSExp
I'll fix it

File: part2/mermaid.ts [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+};
+
+const parseIfExp = (edgeList: M.Edge[], exp: L4.IfExp, head: M.NodeRef, label?: M.EdgeLabel): Result<M.Edge[]> =>
+{
+    const ifExp: M.NodeDecl = M.makeNodeDecl(makeCountGen("IfExp"), "IfExp");
+    const ifRef: M.NodeRef = nodeDeclToRef(ifExp);
+    const ifEdge: M.Edge[] = (head.id === "root_0") ? [] : [M.makeEdge(head, ifExp, label)];
+    return bind(mapExpToEdges(edgeList, exp.test, ifRef, M.makeEdgeLabel("test")),
+                (testEdges: M.Edge[]) => bind(mapExpToEdges(testEdges, exp.then, ifRef, M.makeEdgeLabel("than")),
+                                                (thenEdges: M.Edge[]) => bind(mapExpToEdges(thenEdges, exp.alt, ifRef, M.makeEdgeLabel("alt")),
```
Comment (omerporze) [inline part2/mermaid.ts:112]: can you put them in same indent?

File: part2/mermaid.ts [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+};
+
+const parseAppExp = (edgeList: M.Edge[], exp: L4.AppExp, head: M.NodeRef, label?: M.EdgeLabel): Result<M.Edge[]> =>
+{
+    const app: M.NodeDecl = M.makeNodeDecl(makeCountGen("AppExp"), "AppExp");
+    const appRef: M.NodeRef = nodeDeclToRef(app);
+    const appEdge: M.Edge[] = (head.id === "root_0") ? [] : [M.makeEdge(head, app, label)];
+    return bind(mapExpToEdges(edgeList, exp.rator, appRef, M.makeEdgeLabel("rator")),
+                (ratorEdges: M.Edge[]) => bind(parseArray(ratorEdges, exp.rands, appRef, M.makeEdgeLabel("rands")),
+                                                (randsEdges: M.Edge[]) => makeOk(flatten([appEdge, randsEdges]))
```
Comment (omerporze) [inline part2/mermaid.ts:100]: can you put them in same indent?

File: part2/mermaid.ts [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+};
+
+const parseLetExp = (edgeList: M.Edge[], exp: L4.LetExp, head: M.NodeRef, label?: M.EdgeLabel): Result<M.Edge[]> =>
+{
+    const letExp: M.NodeDecl = M.makeNodeDecl(makeCountGen("LetExp"), "LetExp");
+    const letRef: M.NodeRef = nodeDeclToRef(letExp);
+    const letEdge: M.Edge[] = (head.id === "root_0") ? [] : [M.makeEdge(head, letExp, label)];
+    return bind(parseArray(edgeList, exp.body, letRef, M.makeEdgeLabel("body")),
+                (bodyEdge: M.Edge[]) => bind(parseBindings(bodyEdge, exp.bindings, letRef),
+                                            (bindingsEdges: M.Edge[]) => makeOk(flatten([letEdge, bindingsEdges]))
```
Comment (omerporze) [inline part2/mermaid.ts:156]: can you put them in same indent?

File: part2/mermaid.ts [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+const parseLetrecExp = (edgeList: M.Edge[], exp: L4.LetrecExp, head: M.NodeRef, label?: M.EdgeLabel): Result<M.Edge[]> =>
+{
+    const letExp: M.NodeDecl = M.makeNodeDecl(makeCountGen("LetrecExp"), "LetrecExp");
+    const letRef: M.NodeRef = nodeDeclToRef(letExp);
+    const letEdge: M.Edge[] = (head.id === "root_0") ? [] : [M.makeEdge(head, letExp, label)];
+    return bind(parseArray(edgeList, exp.body, letRef, M.makeEdgeLabel("body")),
+                (bodyEdge: M.Edge[]) => bind(parseBindings(bodyEdge, exp.bindings, letRef),
+                                            (bindingsEdges: M.Edge[]) => makeOk(flatten([letEdge, bindingsEdges]))
+                                            )
```
Comment (omerporze) [inline part2/mermaid.ts:169]: can you put them in same indent?

File: part2/mermaid.ts [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+const parseCompoundSexp = (exp: CompoundSExp, head: M.NodeRef, label?: M.EdgeLabel): Result<M.Edge[]> =>
+{
+    const compundSexp: M.NodeDecl = M.makeNodeDecl(makeCountGen("CompoundSexp"), "CompoundSexp");
+    const compundSexpRef: M.NodeRef = nodeDeclToRef(compundSexp);
+    const compundSexpEdge: M.Edge[] = (head.id === "root_0") ? [] : [M.makeEdge(head, compundSexp, label)];
+    return bind(parseSexpValue(exp.val1, compundSexpRef, M.makeEdgeLabel("val1")),
+                (val1Edge: M.Edge[]) => bind(parseSexpValue(exp.val2, compundSexpRef, M.makeEdgeLabel("val2")),
+                                        (val2Edge: M.Edge[]) => makeOk(flatten([compundSexpEdge, val1Edge, val2Edge]))
+                                            )
```
Comment (omerporze) [inline part2/mermaid.ts:181]: can you put them in same indent?

**Pull request merged**